process_scheduling_algorithms/preemptive_priority_scheduling.c

```diff
@@ -0,0 +1,158 @@
+ */
+#include <stdio.h>    /// for IO operations (`printf`)
+/**
+ * @brief Structure to represent a process
+ */
+struct Process {
+    int id;             /**< Process ID */
+    int arrival_time;   /**< Arrival time of the process */
+    int burst_time;     /**< Burst time of the process */
+    int priority;       /**< Priority of the process */
+    int executed_time;  /**< Time the process has been executed */
+    int completion_time;/**< Completion time of the process */
+};
+/**
+ * @brief Performs preemptive priority scheduling for a set of processes.
+ * @param processes An array of Process structures representing the processes to be scheduled.
+ * @param n The number of processes in the array.
+ * @return void
+ */
+void preemptivePriority(struct Process processes[], int n) {
+    int currentTime = 0;
+    int executedProcesses = 0;
+    int currentProcess = -1;
+    float totalTurnaroundTime = 0;
+    float totalWaitingTime = 0;
+
+    while (executedProcesses < n) {
+        int highestPriority = 9999; // A large number as initial priority
+        int nextProcess = -1;
+
+        for (int i = 0; i < n; i++) {
+            if (processes[i].arrival_time <= currentTime && processes[i].executed_time < processes[i].burst_time) {
+                if (processes[i].priority < highestPriority) {
+                    highestPriority = processes[i].priority;
+                    nextProcess = i;
+                }
+            }
+        }
+
+        if (nextProcess != -1) {
+            if (currentProcess != -1 && currentProcess != nextProcess) {
+                printf("Time %d: Process %d -> Process %d\n", currentTime, currentProcess + 1, nextProcess + 1);
+            }
+            currentProcess = nextProcess;
+            processes[currentProcess].executed_time++;
+            currentTime++;
+
+            if (processes[currentProcess].executed_time == processes[currentProcess].burst_time) {
+                executedProcesses++;
+                processes[currentProcess].completion_time = currentTime;
+                printf("Time %d: Process %d completed\n", currentTime, currentProcess + 1);
+                totalTurnaroundTime += (float)processes[currentProcess].completion_time - processes[currentProcess].arrival_time;
+                totalWaitingTime += (float)processes[currentProcess].completion_time - processes[currentProcess].arrival_time - processes[currentProcess].burst_time;
+                currentProcess = -1;
+            }
+        } else {
+            currentTime++;
+        }
+    }
+
+    float avgTurnaroundTime = totalTurnaroundTime / n;
+    float avgWaitingTime = totalWaitingTime / n;
+
+    printf("\nAverage Turnaround Time: %.2f\n", avgTurnaroundTime);
+    printf("Average Waiting Time: %.2f\n", avgWaitingTime);
+}
+/**
+ * @brief Runs test cases for preemptive priority scheduling.
+ * @return void
+ */
+static void test() {
+    struct Process testCase1[] = {
```

Comment:
Provide negative test cases or use unsigned inputs

Reply:
I have used unsigned input. Thanks for the suggestion